### ISSUE TITLE
Research on Apache Spark on how to do sentiment analysis (2/6/2025)

### ISSUE THREAD

**Issue by jaslene-u** [Jingyong14/Jingyong14#74]
- Watch tutorial provided by Dr.Aryati  [YouTube Link](https://www.youtube.com/watch?v=D4Z-r69Xj2Q)
- Research regarding the installation and environment variable setting to initialize Apache Spark
- More details can refers this link: [Jaslene's Lab2](https://docs.google.com/document/d/1LwkMdTAQMotkIT4QxpCEGKQoK2MkPFb89x_sGy07M6E/edit?usp=sharing)
- Created a template of sentiment analysis using PySpark to be used in the future

###  Install module needed

from pyspark.sql import SparkSession
from pyspark.sql.functions import col, lower, regexp_replace, trim
from pyspark.ml import Pipeline
from pyspark.ml.feature import Tokenizer, StopWordsRemover, HashingTF, IDF
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

#### 1. Initialize SparkSession

spark = SparkSession.builder \
    .appName("SentimentAnalysisFromCSV") \
    .getOrCreate()

print("✅ Spark session created successfully.")

#### 2. Load Data from CSV

csv_path = "your_dataset.csv"  # Replace with your actual path
df = spark.read.csv(csv_path, header=True, inferSchema=True)

print("✅ CSV file loaded successfully.")
df.printSchema()
df.show(5)

#### 3. Clean Text Column

cleaned_df = (
    df.withColumn("text", lower(col("text")))
      .withColumn("text", regexp_replace(col("text"), r"http\S+|@\S+|#\S+", ""))
      .withColumn("text", regexp_replace(col("text"), r"[^a-zA-Z\s]", ""))
      .withColumn("text", trim(col("text")))
)

print("✅ Text cleaning completed.")

#### 4. Split Data into Train and Test Sets

train_df, test_df = cleaned_df.randomSplit([0.8, 0.2], seed=42)
print(f"✅ Data split into training ({train_df.count()} records) and testing ({test_df.count()} records).")

#### 5. Set Up ML Pipeline

tokenizer = Tokenizer(inputCol="text", outputCol="words")
remover = StopWordsRemover(inputCol="words", outputCol="filtered_words")
hashingTF = HashingTF(inputCol="filtered_words", outputCol="rawFeatures", numFeatures=1000)
idf = IDF(inputCol="rawFeatures", outputCol="features")
classifier = LogisticRegression(featuresCol="features", labelCol="label")

pipeline = Pipeline(stages=[tokenizer, remover, hashingTF, idf, classifier])
print("✅ ML pipeline created.")

#### 6. Train the Model

model = pipeline.fit(train_df)
print("✅ Model training completed on training data.")

#### 7. Make Predictions on Test Set

predictions = model.transform(test_df)
print("✅ Predictions made on test data.")
predictions.select("text", "label", "prediction").show(10, truncate=False)

#### 8. Evaluate the Model

evaluator_accuracy = MulticlassClassificationEvaluator(labelCol="label", predictionCol="prediction", metricName="accuracy")
evaluator_precision = MulticlassClassificationEvaluator(labelCol="label", predictionCol="prediction", metricName="weightedPrecision")
evaluator_f1 = MulticlassClassificationEvaluator(labelCol="label", predictionCol="prediction", metricName="f1")

accuracy = evaluator_accuracy.evaluate(predictions)
precision = evaluator_precision.evaluate(predictions)
f1 = evaluator_f1.evaluate(predictions)

print(f"📊 Evaluation Results:")
print(f"    - Accuracy : {accuracy:.4f}")
print(f"    - Precision: {precision:.4f}")
print(f"    - F1 Score : {f1:.4f}")

#### 9. Stop Spark

spark.stop()
print("✅ Spark session stopped.")
